File: ws-server/src/main/java/net/devopssolutions/demo/ws/server/component/WsServer.java
package net.devopssolutions.demo.ws.server.component;

import com.fasterxml.jackson.databind.JsonNode;
import com.fasterxml.jackson.databind.ObjectMapper;
import lombok.extern.slf4j.Slf4j;
import net.devopssolutions.demo.ws.server.util.LoggingThreadPoolExecutor;
import org.springframework.beans.factory.annotation.Autowired;
import org.springframework.scheduling.annotation.Scheduled;
import org.springframework.stereotype.Component;

import javax.websocket.*;
import java.io.IOException;
import java.io.InputStream;
import java.nio.ByteBuffer;
import java.security.Principal;
import java.util.Collection;
import java.util.Set;
import java.util.concurrent.ArrayBlockingQueue;
import java.util.concurrent.ConcurrentHashMap;
import java.util.concurrent.TimeUnit;
import java.util.stream.Collectors;
import java.util.zip.GZIPInputStream;

@Slf4j
@Component
public class WsServer extends Endpoint {

    private final ByteBuffer ping = ByteBuffer.allocate(0);
    private final ConcurrentHashMap<String, Session> sessions = new ConcurrentHashMap<>();
    private final LoggingThreadPoolExecutor handlersExecutor = new LoggingThreadPoolExecutor(10, 20, 2, TimeUnit.MINUTES, new ArrayBlockingQueue<>(10_000));

    @Autowired
    private ObjectMapper objectMapper;

    @Autowired
    private RpcMethodDispatcher rpcMethodDispatcher;

    public Set<Session> getAllSessions() {
        return this.sessions.values()
                .stream()
                .map(Session::getOpenSessions)
                .flatMap(Collection::stream)
                .collect(Collectors.toSet());
    }

    public void onBinaryMessage(InputStream message, Session session) {
        log.info("onBinaryMessage id: {}, message: {}", session.getId(), message);
        Principal userPrincipal = session.getUserPrincipal();
        try (GZIPInputStream inputStream = new GZIPInputStream(message)) {
            JsonNode node = objectMapper.readTree(inputStream);
            handlersExecutor.execute(() -> {
                dispatchMessage(node, userPrincipal);
            });
        } catch (Exception e) {
            log.warn("exception handling ws message session: " + session.getId(), e);
        }
    }

    public void onTextMessage(String message, Session session) {
        log.info("onTextMessage id: {}, message: {}", session.getId(), message);
        handlersExecutor.execute(() -> {
            try {
                JsonNode node = objectMapper.readTree(message);
                dispatchMessage(node, session.getUserPrincipal());
            } catch (Exception e) {
                log.warn("exception handling ws message session: " + session.getId(), e);
            }
        });
    }

    private void dispatchMessage(JsonNode node, Principal principal) {
        String id = node.get("id").asText();
        String method = node.get("method").asText();
        JsonNode params = node.get("params");
        rpcMethodDispatcher.handle(id, method, params, principal);
    }

    public void onPong(PongMessage message, Session session) {
        log.info("onPong id: {}, message: {}", session.getId(), message);
    }

    @Override
    public void onOpen(Session session, EndpointConfig config) {
        log.info("onOpen id: {}", session.getId());

        session.setMaxIdleTimeout(10_000);
        session.setMaxBinaryMessageBufferSize(100_000_000);
        session.setMaxTextMessageBufferSize(1_000_000);
        session.addMessageHandler(PongMessage.class, message -> onPong(message, session));
        session.addMessageHandler(String.class, message -> onTextMessage(message, session));
        session.addMessageHandler(InputStream.class, message -> onBinaryMessage(message, session));

        sessions.put(session.getId(), session);
    }

    @Override
    public void onClose(Session session, CloseReason closeReason) {
        log.info("onClose id: {} closeReason: {}", session.getId(), closeReason);
        sessions.remove(session.getId(), session);
    }

    @Override
    public void onError(Session session, Throwable exception) {
        log.info("onError id: " + session.getId(), exception);
        sessions.remove(session.getId(), session);
        try {
            session.close(new CloseReason(CloseReason.CloseCodes.CLOSED_ABNORMALLY, exception.getMessage()));
        } catch (IOException e) {
            log.warn("exception closing session id: " + session.getId(), e);
        }
    }

    @Scheduled(fixedRate = 10_000)
    public void sendPings() {
        log.info("sending pings: {}", sessions.size());
        getAllSessions()
                .forEach(this::sendPing);
    }

    private void sendPing(Session session) {
        log.info("sending ping sessionId: {}, session {}", session.getId(), session);
        try {
            session.getBasicRemote().sendPing(ping);
        } catch (Exception e) {
            boolean remove = sessions.remove(session.getId(), session);
            log.warn("exception sending ping, remove session " + remove, e);
        }
    }

}

File: ws-server/src/main/java/net/devopssolutions/demo/ws/server/component/WsServer.kt
//package net.devopssolutions.demo.ws.server.component
//
//import com.fasterxml.jackson.databind.JsonNode
//import com.fasterxml.jackson.databind.ObjectMapper
//import mu.KLogging
//import net.devopssolutions.demo.ws.model.User
//import net.devopssolutions.demo.ws.server.util.LoggingThreadPoolExecutor
//import org.springframework.scheduling.annotation.Scheduled
//import org.springframework.stereotype.Component
//import org.springframework.web.reactive.socket.WebSocketSession
//import reactor.core.publisher.Mono
//import java.io.IOException
//import java.io.InputStream
//import java.nio.ByteBuffer
//import java.security.Principal
//import java.util.concurrent.ArrayBlockingQueue
//import java.util.concurrent.ConcurrentHashMap
//import java.util.concurrent.TimeUnit
//import java.util.zip.GZIPInputStream
//
//@Component
//open class WsServer(
//        private val rpcMethodDispatcher: RpcMethodDispatcher,
//        private val objectMapper: ObjectMapper
//) {
//    companion object : KLogging()
//
//    private val ping = ByteBuffer.allocate(0)
//    private val sessions = ConcurrentHashMap<String, WebSocketSession>()
//    private val handlersExecutor = LoggingThreadPoolExecutor(10, 20, 2, TimeUnit.MINUTES, ArrayBlockingQueue<Runnable>(200))
//
//    fun getAllSessions(): Collection<WebSocketSession> = sessions.values.toHashSet()
//
//    fun getSession(id: String): WebSocketSession? {
//        return sessions[id]
//    }
//
//    fun onBinaryMessage(message: InputStream, session: WebSocketSession) {
//        logger.info("onBinaryMessage id: {}, message: {}", session.id, message)
//        try {
//            val node = objectMapper.readTree(GZIPInputStream(message))
//            handlersExecutor.execute { dispatchMessage(session.id, node, session.handshakeInfo.principal) }
//        } catch (e: Exception) {
//            logger.warn("exception handling ws message session: " + session.id, e)
//        }
//    }
//
//    fun onTextMessage(message: String, session: WebSocketSession) {
//        logger.info("onTextMessage id: {}, message: {}", session.id, message)
//        handlersExecutor.execute {
//            try {
//                val node = objectMapper.readTree(message)
//                dispatchMessage(session.id, node, session.handshakeInfo.principal)
//            } catch (e: Exception) {
//                logger.warn("exception handling ws message session: " + session.id, e)
//            }
//        }
//    }
//
//    private fun dispatchMessage(sessionId: String, node: JsonNode, principal: Mono<Principal>) {
//        val id = node.get("id").asText()
//        val method = node.get("method").asText()
//        val params = node.get("params")
//        rpcMethodDispatcher.handle(sessionId, id, method, params, principal ?: User("anonymous"))
//    }
//
//    fun onPong(message: PongMessage, session: WebSocketSession) {
//        logger.info("onPong id: {}, message: {}", session.id, message)
//    }
//
//    override fun onOpen(session: WebSocketSession, config: EndpointConfig) {
//        logger.info("onOpen id: {}", session.id)
//
//        session.maxIdleTimeout = 10000
//        session.maxBinaryMessageBufferSize = 100000000
//        session.maxTextMessageBufferSize = 1000000
//        session.addMessageHandler(PongMessage::class.java) { message -> onPong(message, session) }
//        session.addMessageHandler(String::class.java) { message -> onTextMessage(message, session) }
//        session.addMessageHandler(InputStream::class.java) { message -> onBinaryMessage(message, session) }
//
//        sessions.put(session.id, session)
//    }
//
//    fun onClose(session: WebSocketSession, closeReason: CloseReason) {
//        logger.info("onClose id: {} closeReason: {}", session.id, closeReason)
//        sessions.remove(session.id, session)
//    }
//
//    fun onError(session: WebSocketSession, exception: Throwable) {
//        logger.info("onError id: " + session.id, exception)
//        sessions.remove(session.id, session)
//        try {
//            session.close(CloseReason(CloseReason.CloseCodes.CLOSED_ABNORMALLY, exception.message))
//        } catch (e: IOException) {
//            logger.warn("exception closing session id: " + session.id, e)
//        }
//    }
//
//    @Scheduled(fixedRate = 10_000)
//    fun sendPings() {
//        logger.info("sending pings: {}", sessions.size)
//        getAllSessions().forEach { this.sendPing(it) }
//    }
//
//    private fun sendPing(session: WebSocketSession) {
//        logger.info("sending ping sessionId: {}, session {}", session.id, session)
//        try {
//            session.basicRemote.sendPing(ping)
//        } catch (e: Exception) {
//            val remove = sessions.remove(session.id, session)
//            logger.warn("exception sending ping, remove session " + remove, e)
//        }
//    }
//
//}
